google/cloud/internal/completion_queue_impl.cc

```diff
@@ -69,8 +153,39 @@ void CompletionQueueImpl::CancelAll() {
   }
 }
 
-std::unique_ptr<grpc::Alarm> CompletionQueueImpl::CreateAlarm() const {
-  return absl::make_unique<grpc::Alarm>();
+future<StatusOr<std::chrono::system_clock::time_point>>
+CompletionQueueImpl::MakeDeadlineTimer(
+    std::chrono::system_clock::time_point deadline) {
+  auto op =
+      std::make_shared<AsyncTimerFuture>(absl::make_unique<grpc::Alarm>());
+  StartOperation(op, [&](void* tag) { op->Set(cq(), deadline, tag); });
+  return op->GetFuture();
+}
+
+void CompletionQueueImpl::RunAsync(
+    std::unique_ptr<internal::RunAsyncBase> function) {
+  auto op = std::make_shared<AsyncFunction>(std::move(function));
+  StartOperation(op, [&](void* tag) { op->Set(cq(), tag); });
+}
+
+void CompletionQueueImpl::StartOperation(std::shared_ptr<AsyncGrpcOperation> op,
+                                         absl::FunctionRef<void(void*)> start) {
+  void* tag = op.get();
+  std::unique_lock<std::mutex> lk(mu_);
+  if (shutdown_) {
+    lk.unlock();
+    op->Notify(/*ok=*/false);
+    return;
+  }
+  auto ins =
+      pending_ops_.emplace(reinterpret_cast<std::intptr_t>(tag), std::move(op));
+  if (ins.second) {
+    start(tag);
+    lk.unlock();
+    return;
+  }
+  google::cloud::internal::ThrowRuntimeError(
+      "assertion failure: insertion should succeed");
```

Comment:
Perhaps something involving "duplicate tag" would be helpful?

Reply:
Done.